lib/web_ui/dev/roll_fallback_fonts.dart

```diff
@@ -509,3 +494,354 @@ bool _checkForLicenseAttribution(Uint8List fontBytes) {
   }
   return false;
 }
+
+class _Font {
+  _Font(this.family, this.index, this.starts, this.ends);
+
+  final String family;
+  final int index;
+  final List<int> starts;
+  final List<int> ends; // inclusive ends
+
+  static int compare(_Font a, _Font b) => a.index.compareTo(b.index);
+
+  String get shortName =>
+      _shortName +
+      String.fromCharCodes(
+          '$index'.codeUnits.map((int ch) => ch - 48 + 0x2080));
+
+  String get _shortName => family.startsWith('Noto Sans ')
+      ? family.substring('Noto Sans '.length)
+      : family;
+}
+
+/// The boundary of a range of a font.
+class _Boundary {
+  _Boundary(this.value, this.isStart, this.font);
+  final int value; // inclusive start or exclusive end.
+  final bool isStart;
+  final _Font font;
+
+  static int compare(_Boundary a, _Boundary b) => a.value.compareTo(b.value);
+}
+
+class _Range {
+  _Range(this.start, this.end, this.fontSet);
+  final int start;
+  final int end;
+  final _FontSet fontSet;
+
+  @override
+  String toString() {
+    return '[${start.toRadixString(16)}, ${end.toRadixString(16)}]'
+        ' (${end - start + 1})'
+        ' ${fontSet.description()}';
+  }
+}
+
+/// A canonical representative for a set of _Fonts. The fonts are stored in
+/// order of increasing `_Font.index`.
+class _FontSet {
+  _FontSet(this.fonts);
+
+  /// The number of [_Font]s in this set.
+  int get length => fonts.length;
+
+  /// The members of this set.
+  final List<_Font> fonts;
+
+  /// Number of unicode ranges that are supported by this set of fonts.
+  int rangeCount = 0;
+
+  /// The serialization order of this set. This index is assigned after building
+  /// all the sets.
+  late final int index;
+
+  static int orderByDecreasingRangeCount(_FontSet a, _FontSet b) {
+    final int r = b.rangeCount.compareTo(a.rangeCount);
+    if (r != 0) {
+      return r;
+    }
+    return orderByLexicographicFontIndexes(a, b);
+  }
+
+  static int orderByLexicographicFontIndexes(_FontSet a, _FontSet b) {
+    for (int i = 0; i < a.length && i < b.length; i++) {
+      final int r = _Font.compare(a.fonts[i], b.fonts[i]);
+      if (r != 0) {
+        return r;
+      }
+    }
+    assert(a.length != b.length); // _FontSets are canonical.
+    return a.length - b.length;
+  }
+
+  @override
+  String toString() {
+    return description();
+  }
+
+  String description() {
+    return fonts.map((_Font font) => font.shortName).join(', ');
+  }
+}
+
+/// A trie node [1] used to find the canonical _FontSet.
+///
+/// [1]: https://en.wikipedia.org/wiki/Trie
+class _TrieNode {
+  final Map<_Font, _TrieNode> _children = <_Font, _TrieNode>{};
+  _FontSet? fontSet;
+
+  _TrieNode insertAtRoot(Iterable<_Font> fonts) {
```

Comment:
IIUC this adds the fonts (eg `[font1, font2, font3, ...]`) in a line, not all as children of the Trie that `insertAtRoot` is called on, which is what I assumed when I read the function name. Add docs explaining what this does and what it returns?

```
font1
  \
  font2
    \
    font3
    ...
```

Reply:
I have renamed the function and added a comment to clarify that it follows a path trough the trie.